src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+package com.example.onlinebookstore.service.impl;
+
+import com.example.onlinebookstore.dto.shoppingcart.CartItemDto;
+import com.example.onlinebookstore.dto.shoppingcart.ShoppingCartDto;
+import com.example.onlinebookstore.exception.EntityNotFoundException;
+import com.example.onlinebookstore.mapper.ShoppingCartMapper;
+import com.example.onlinebookstore.model.Book;
+import com.example.onlinebookstore.model.CartItem;
+import com.example.onlinebookstore.model.ShoppingCart;
+import com.example.onlinebookstore.model.User;
+import com.example.onlinebookstore.repository.BookRepository;
+import com.example.onlinebookstore.repository.CartItemRepository;
+import com.example.onlinebookstore.repository.ShoppingCartRepository;
+import com.example.onlinebookstore.repository.UserRepository;
+import com.example.onlinebookstore.service.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Transactional
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        ShoppingCart shoppingCart = getOrCreateShoppingCart(userId);
+        return shoppingCartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartDto addCartItem(Long userId, CartItemDto cartItemDto) {
+        Book book = bookRepository.findById(cartItemDto.bookId()).orElseThrow(
+                () -> new EntityNotFoundException("The book is not found"));
+        ShoppingCart shoppingCart = getOrCreateShoppingCart(userId);
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(book);
+        cartItem.setQuantity(cartItemDto.quantity());
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartDto updateCartItem(Long cartItemId, Integer quantity) {
+        CartItem cartItem = cartItemRepository
+                .findById(cartItemId)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Cart item with given id is not found"));
+        cartItem.setQuantity(quantity);
+        return shoppingCartMapper.toShoppingCartDto(cartItem.getShoppingCart());
+    }
+
+    @Override
+    public void deleteCartItem(Long id) {
+        cartItemRepository.deleteById(id);
+    }
```

Comment:
I'm not sure about how this method works in your project. Here's what I think: a delete method for a cart item is basically an update method for a shopping cart. I think it would make more sense if did not give void type to this method but returned the updated ShoppingCartDto without the deleted cart item. Think about this, but I'll just leave it for mentors to decide